File: windows/src/SSMConnect.Domain/DcvConnectionFile.cs
using System.Globalization;
using System.Text;

namespace SSMConnect.Domain;

/// <summary>
/// An Amazon DCV connection file (<c>.dcv</c>, INI format) used to auto-login the viewer.
/// </summary>
/// <remarks>
/// Rendering lives in the domain because both clients must produce byte-identical content; writing,
/// permissioning, launching, and deleting the file are platform concerns and live in the adapter.
/// <para>
/// The host is the IPv4 loopback literal, never <c>localhost</c>. The SSM port-forward binds IPv4
/// only, while <c>localhost</c> can resolve to <c>::1</c> first — the viewer then connects to an
/// address with no listener and reports the endpoint as unreachable.
/// </para>
/// </remarks>
public sealed record DcvConnectionFile
{
    public const string TempFilePrefix = "ssm-connect-";
    public const string FileExtension = "dcv";
    public const string DefaultUser = "ec2-user";
    public const string LoopbackHost = "127.0.0.1";

    public string Host { get; init; } = LoopbackHost;
    public required int Port { get; init; }
    public string User { get; init; } = DefaultUser;

    /// <summary>Single-user auth: the Secrets Manager password. Null in multi-user mode.</summary>
    public string? Password { get; init; }

    /// <summary>Multi-user auth: a presigned identity token. Null in single-user mode.</summary>
    public string? AuthToken { get; init; }

    /// <summary>The user's own virtual session. Null targets the default session.</summary>
    public string? SessionId { get; init; }

    public string WebUrlPath { get; init; } = "/";

    /// <summary>Single-user connection: a shared user plus a Secrets Manager password.</summary>
    public static DcvConnectionFile SingleUser(int port, string password, string user = DefaultUser) =>
        new() { Port = port, User = user, Password = password };

    /// <summary>Multi-user connection: the caller's own session, authorised by a presigned token.</summary>
    public static DcvConnectionFile MultiUser(int port, string user, string sessionId, string authToken) =>
        new() { Port = port, User = user, SessionId = sessionId, AuthToken = authToken };

    /// <summary>
    /// Renders the INI content the viewer expects, emitting only the fields relevant to the active
    /// auth mode.
    /// </summary>
    public string ToIni()
    {
        var builder = new StringBuilder();
        builder.Append("[version]\n");
        builder.Append("format=1.0\n");
        builder.Append('\n');
        builder.Append("[connect]\n");
        builder.Append(CultureInfo.InvariantCulture, $"host={Host}\n");
        builder.Append(CultureInfo.InvariantCulture, $"port={Port.ToString(CultureInfo.InvariantCulture)}\n");
        builder.Append(CultureInfo.InvariantCulture, $"user={User}\n");
        if (SessionId is not null)
        {
            builder.Append(CultureInfo.InvariantCulture, $"sessionid={SessionId}\n");
        }

        if (AuthToken is not null)
        {
            builder.Append(CultureInfo.InvariantCulture, $"authtoken={AuthToken}\n");
        }

        if (Password is not null)
        {
            builder.Append(CultureInfo.InvariantCulture, $"password={Password}\n");
        }

        builder.Append(CultureInfo.InvariantCulture, $"weburlpath={WebUrlPath}");
        return builder.ToString();
    }
}

File: SSMConnectKit/Sources/SSMConnectDomain/DCVConnectionFile.swift
import Foundation

/// An Amazon DCV **connection file** (`.dcv`, INI format) used to auto-login the DCV Viewer (ADR-8).
///
/// Opening a populated connection file makes DCV Viewer connect and authenticate without manual
/// host/password entry — the only DCV automation hook for credential injection. The secret is
/// injected here transiently: the file is written `0600`, opened, and deleted immediately (F-10).
///
/// Two auth modes are supported:
/// - **Vanilla (single-user):** `password=` from Secrets Manager, `user=ec2-user`. Unchanged from v1.
/// - **Multi-user:** `authtoken=` (a presigned-identity token) + `sessionid=` targeting the user's
///   own virtual session. No password. The native client supports both `authtoken` and `sessionid`
///   in the `[connect]` section.
public struct DCVConnectionFile: Equatable, Sendable {
    public init(
        host: String = "127.0.0.1",
        port: Int,
        user: String = "ec2-user",
        password: String? = nil,
        authToken: String? = nil,
        sessionId: String? = nil,
        webUrlPath: String = "/"
    ) {
        self.host = host
        self.port = port
        self.user = user
        self.password = password
        self.authToken = authToken
        self.sessionId = sessionId
        self.webUrlPath = webUrlPath
    }

    /// IPv4 loopback (not `localhost`): the SSM port-forward binds IPv4 `127.0.0.1` only, while
    /// `localhost` resolves to IPv6 `::1` first on macOS — the DCV Viewer then connects to `::1`,
    /// finds no listener, and fails with "endpoint is unreachable" (see docs/specs/bug-ipv6-localhost-*).
    public var host: String = "127.0.0.1"
    public var port: Int
    public var user: String = "ec2-user"
    /// Vanilla auth: Secrets-Manager password. `nil` in multi-user mode.
    public var password: String?
    /// Multi-user auth: presigned-identity token. `nil` in vanilla mode.
    public var authToken: String?
    /// Target DCV session id (the user's own virtual session). `nil` for the default session.
    public var sessionId: String?
    /// DCV web URL path; `/` for a default session.
    public var webUrlPath: String = "/"

    /// Temp-file naming so the startup sweep can find orphans left by a crash (ADR-8, §8).
    public static let tempFilePrefix = "ssm-connect-"
    public static let fileExtension = "dcv"

    /// Vanilla (single-user) connection: `user` + Secrets-Manager `password`.
    public static func vanilla(host: String = "127.0.0.1", port: Int, user: String = "ec2-user", password: String) -> DCVConnectionFile {
        DCVConnectionFile(host: host, port: port, user: user, password: password)
    }

    /// Multi-user connection: `user`'s own `sessionId`, authorized by a presigned-identity `authToken`.
    public static func multiUser(host: String = "127.0.0.1", port: Int, user: String, sessionId: String, authToken: String) -> DCVConnectionFile {
        DCVConnectionFile(host: host, port: port, user: user, password: nil, authToken: authToken, sessionId: sessionId)
    }

    /// Renders the INI content DCV Viewer expects. Emits only the fields relevant to the active
    /// auth mode (token fields when present, otherwise the password field).
    public func iniContent() -> String {
        var lines = [
            "[version]",
            "format=1.0",
            "",
            "[connect]",
            "host=\(host)",
            "port=\(port)",
            "user=\(user)",
        ]
        if let sessionId { lines.append("sessionid=\(sessionId)") }
        if let authToken { lines.append("authtoken=\(authToken)") }
        if let password { lines.append("password=\(password)") }
        lines.append("weburlpath=\(webUrlPath)")
        return lines.joined(separator: "\n")
    }
}
